File: src/libbitcoinpqc/nodejs/examples/basic-usage.ts
import {
  Algorithm,
  generateKeyPair,
  sign,
  verify,
  publicKeySize,
  secretKeySize,
  signatureSize,
} from "../src";
import crypto from "crypto";

/**
 * This example demonstrates basic usage of the bitcoinpqc TypeScript bindings.
 *
 * It shows:
 * 1. Getting key and signature sizes
 * 2. Key generation
 * 3. Signing messages
 * 4. Verifying signatures
 * 5. Error handling
 */

// Print key sizes for all algorithms
console.log("===== Key and Signature Sizes =====");
for (const algo of [
  Algorithm.ML_DSA_44,
  Algorithm.SLH_DSA_SHAKE_128S,
  Algorithm.FN_DSA_512,
]) {
  const algoName = Algorithm[algo];
  console.log(`${algoName}:`);
  console.log(`  Public key size: ${publicKeySize(algo)} bytes`);
  console.log(`  Secret key size: ${secretKeySize(algo)} bytes`);
  console.log(`  Signature size: ${signatureSize(algo)} bytes`);
  console.log();
}

// Generate a keypair
function generateKeysAndSign(algorithm: Algorithm, message: string): void {
  console.log(`===== Working with ${Algorithm[algorithm]} =====`);

  // Generate random data for key generation
  console.log("Generating random data...");
  const randomData = crypto.randomBytes(128);

  try {
    // Generate a keypair
    console.log("Generating keypair...");
    const keypair = generateKeyPair(algorithm, randomData);
    console.log(
      `Generated keypair with public key size ${keypair.publicKey.bytes.length} bytes`
    );

    // Create a message to sign
    const messageBytes = Buffer.from(message, "utf-8");
    console.log(`Message to sign: "${message}"`);

    // Sign the message
    console.log("Signing message...");
    const signature = sign(keypair.secretKey, messageBytes);
    console.log(`Created signature of size ${signature.bytes.length} bytes`);

    // Verify the signature
    console.log("Verifying signature...");
    try {
      verify(keypair.publicKey, messageBytes, signature);
      console.log("✅ Signature verified successfully!");
    } catch (error) {
      console.error("❌ Signature verification failed:", error);
    }

    // Try verifying with a different message (should fail with real implementation)
    const badMessage = Buffer.from(message + " (modified)", "utf-8");
    console.log(
      `\nAttempting to verify with modified message: "${message} (modified)"`
    );
    try {
      verify(keypair.publicKey, badMessage, signature);
      console.log("❌ Verification succeeded with modified message!");
    } catch (error) {
      console.log("✅ Verification correctly failed with modified message");
      console.log(`   Error: ${error}`);
    }
  } catch (error) {
    console.error(
      `❌ Error while working with ${Algorithm[algorithm]}:`,
      error
    );
  }

  console.log("\n");
}

// Run examples with working algorithms
generateKeysAndSign(Algorithm.FN_DSA_512, "Hello from FN-DSA-512 (Falcon)!");
generateKeysAndSign(Algorithm.ML_DSA_44, "Hello from ML-DSA-44 (Dilithium)!");
generateKeysAndSign(
  Algorithm.SLH_DSA_SHAKE_128S,
  "Hello from SLH-DSA-SHAKE-128S (SPHINCS+)!"
);

File: src/libbitcoinpqc/python/examples/basic_usage.py
#!/usr/bin/env python3
"""
Basic usage example for the bitcoinpqc Python bindings.
"""

import os
import sys
import secrets
from pathlib import Path
import time

# Add the parent directory to the path so we can import the module
sys.path.insert(0, str(Path(__file__).parent.parent))

from bitcoinpqc import Algorithm, keygen, sign, verify


def main():
    """Run a basic example of bitcoinpqc usage."""
    # Choose an algorithm
    algorithm = Algorithm.ML_DSA_44  # CRYSTALS-Dilithium

    # Generate random data for key generation
    random_data = secrets.token_bytes(128)

    print(f"Generating key pair for {algorithm.name}...")
    start = time.time()
    keypair = keygen(algorithm, random_data)
    keygen_time = time.time() - start

    print(f"Key generation took {keygen_time:.4f} seconds")
    print(f"Public key size: {len(keypair.public_key)} bytes")
    print(f"Secret key size: {len(keypair.secret_key)} bytes")

    # Sign a message
    message = b"This is a test message for Bitcoin PQC signatures."
    print(f"\nSigning message: {message.decode('utf-8')}")

    start = time.time()
    signature = sign(algorithm, keypair.secret_key, message)
    sign_time = time.time() - start

    print(f"Signing took {sign_time:.4f} seconds")
    print(f"Signature size: {len(signature.signature)} bytes")

    # Verify the signature
    print("\nVerifying signature...")

    start = time.time()
    is_valid = verify(algorithm, keypair.public_key, message, signature)
    verify_time = time.time() - start

    print(f"Verification took {verify_time:.4f} seconds")
    print(f"Signature is valid: {is_valid}")

    # Verify with incorrect message
    bad_message = b"This is NOT the original message!"
    print("\nVerifying signature with incorrect message...")

    is_valid = verify(algorithm, keypair.public_key, bad_message, signature)
    print(f"Signature is valid (should be False): {is_valid}")


if __name__ == "__main__":
    main()
